controllers/velero.go

```diff
@@ -380,142 +385,105 @@ func (r *VeleroReconciler) buildVeleroDeployment(veleroDeployment *appsv1.Deploy
-			MountPath: "/scratch",
-		},
-		{
+	//check if CSI plugin is added in spec
+	for _, plugin := range velero.Spec.DefaultVeleroPlugins {
+		if plugin == oadpv1alpha1.DefaultPluginCSI {
+			// CSI plugin is added so ensure that CSI feature flags is set
+			foundCSIFeatureFlag := false
+			for _, featureFlag := range velero.Spec.VeleroFeatureFlags {
+				if featureFlag == enableCSIFeatureFlag {
+					foundCSIFeatureFlag = true
+					break
+				}
+			}
+			if !foundCSIFeatureFlag { // Not Found so append to feature flag
+				velero.Spec.VeleroFeatureFlags = append(velero.Spec.VeleroFeatureFlags, enableCSIFeatureFlag)
+			}
+			break
+		}
+	}
+	deploymentName := veleroDeployment.Name //saves desired deployment name before install.Deployment overwrites them.
+	*veleroDeployment = *install.Deployment(veleroDeployment.Namespace,
+		install.WithResources(r.getVeleroResourceReqs(velero)),
+		install.WithImage(getVeleroImage()),
+		install.WithFeatures(velero.Spec.VeleroFeatureFlags),
+		// use WithSecret false even if we have secret because we use a different VolumeMounts and EnvVars
+		// see: https://github.com/vmware-tanzu/velero/blob/ed5809b7fc22f3661eeef10bdcb63f0d74472b76/pkg/install/deployment.go#L223-L261
+		// our secrets are appended to containers/volumeMounts in credentials.AppendPluginSpecificSpecs function
+		install.WithSecret(false),
+	)
+	// adjust veleroDeployment from install
+	return r.customizeVeleroDeployment(velero, veleroDeployment, deploymentName) //reapply saved deploymentName
+}
+
+func (r *VeleroReconciler) customizeVeleroDeployment(velero *oadpv1alpha1.Velero, veleroDeployment *appsv1.Deployment, deploymentName string) error {
+	veleroDeployment.Name = deploymentName
+	veleroDeployment.Labels = r.getAppLabels(velero)
+	veleroDeployment.Spec.Selector = veleroLabelSelector
+
+	//TODO: add velero nodeselector, needs to be added to the VELERO CR first
+	// Selector: veleroDeployment.Spec.Selector,
+	veleroDeployment.Spec.Replicas = pointer.Int32(1)
+	veleroDeployment.Spec.Template.Spec.Tolerations = velero.Spec.VeleroTolerations
+	veleroDeployment.Spec.Template.Spec.Volumes = append(veleroDeployment.Spec.Template.Spec.Volumes,
+		corev1.Volume{
+			Name: "certs",
+			VolumeSource: corev1.VolumeSource{
+				EmptyDir: &corev1.EmptyDirVolumeSource{},
+			},
+		})
+	//add any default init containers here if needed eg: setup-certificate-secret
+	if veleroDeployment.Spec.Template.Spec.InitContainers == nil {
+		veleroDeployment.Spec.Template.Spec.InitContainers = []corev1.Container{}
+	}
+
+	var veleroContainer *corev1.Container
+	for i, container := range veleroDeployment.Spec.Template.Spec.Containers {
+		if container.Name == common.Velero {
+			veleroContainer = &veleroDeployment.Spec.Template.Spec.Containers[i]
+			break
+		}
+	}
+	if err := r.customizeVeleroContainer(velero, veleroDeployment, veleroContainer); err != nil {
+		return err
+	}
+	return credentials.AppendPluginSpecificSpecs(velero, veleroDeployment, veleroContainer)
```

Comment:
@kaovilai When appending plugin specific specs can you please check if the user has specified secret name under Velero CR for the specific BSL and use that as secret name instead of the default secret name. For instance, Under Velero CR -> BSL spec -> Credential -> Name -> `foo-crdentials` for AWS BSL. Then here, instead of using the default name `cloud-credentials` from the map `PluginSpecificFields` use the name `foo-credentials` wherever the secret name is needed (VolumeMounts, Volumes etc) in appending plugin specific specifications to velero deployment as well as restic ds.